dask/layers.py

```diff
@@ -100,6 +103,211 @@ def __init__(
         )
 
 
+class ArrayOverlapLayer(Layer):
+    """Simple HighLevelGraph array overlap layer.
+
+    Lazily computed High-level graph layer for a array overlap operations.
+
+    Parameters
+    ----------
+    name : str
+        Name of new output overlap array.
+    array : Dask array
+    axes: Mapping
+        Axes dictionary indicating overlap in each dimension,
+        e.g. ``{'0': 1, '1': 1}``
+    """
+
+    def __init__(
+        self,
+        name,
+        axes,
+        chunks,
+        numblocks,
+        token,
+    ):
+        super().__init__()
+        self.name = name
+        self.axes = axes
+        self.chunks = chunks
+        self.numblocks = numblocks
+        self.token = token
+        self._cached_keys = None
+
+    def __repr__(self):
+        return "ArrayOverlapLayer<name='{}'".format(self.name)
+
+    @property
+    def _dict(self):
+        """Materialize full dict representation"""
+        if hasattr(self, "_cached_dict"):
+            return self._cached_dict
+        else:
+            dsk = self._construct_graph()
+            self._cached_dict = dsk
+        return self._cached_dict
+
+    def __getitem__(self, key):
+        return self._dict[key]
+
+    def __iter__(self):
+        return iter(self._dict)
+
+    def __len__(self):
+        return len(self._dict)
```

Comment:
@gjoseph92 made an excellent point earlier that `__len__` and `__iter__` will materialize the whole task graph, and this is all too easy to do accidentally (eg: `_repr_html()` calls `__len__`).

When I talked about this with Ian, we were not quite sure how to calculate the number of keys in the dictionary *a priori*. That's very easy to know with `map_blocks`, but harder to work out with `map_overlap`. The number of tasks depends on lots of things, including:
* the number of chunks in the array
* how those chunks are arranged together (is it a line of chunks, is it lots of little squares making up a bigger square)
* the overlap depth in pixels (and this seems to change it in really weird ways)
* the boundary conditions (None, reflect, periodic, nearest, constant)

It's probably not an intractable problem, but it's not straightforward. Any advice is appreciated.

Reply:
I thought I'd work it out if I just sat down and spent a bit of focused time thinking about it, but nope I can't work out the logic.

I find it especially odd that the number of dictionary keys *decreases* like it hits a some kind of a cliff as you increase the overlap depth:

```python
import numpy as np
import dask.array as da

x = da.from_array(np.random.random((10,10)), chunks=(5,5))

result1 = x.map_blocks(lambda x: x)
print(len(result1.dask))
# 8

result2 = x.map_overlap(lambda x: x, depth=2, boundary='None')
print(len(result2.dask))
# 166

result3 = x.map_overlap(lambda x: x, depth=7, boundary='None')
print(len(result3.dask))
# 83
```

Reply:
Even if you can't come up with a perfectly closed-form solution, at the very worst, I bet generating all the keys in an iterator would still be faster than generating the full dict of tasks. It could be "partially closed form".

Reply:
Excited about this work to make overlaps more tractable!

I recently hit the auto-rechunking issue myself.  Is it a feature worth the surprises it generates, like this one?